pandas/_libs/tslibs/ccalendar.pyx

```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
+# cython: profile=False
+# cython: boundscheck=False
+"""
+Cython implementations of functions resembling the stdlib calendar module
+"""
+
+cimport cython
+from cython cimport Py_ssize_t
+
+import numpy as np
+cimport numpy as np
+from numpy cimport int64_t, int32_t
+np.import_array()
+
+
+# ----------------------------------------------------------------------
+# Constants
+
+# Slightly more performant cython lookups than a 2D table
+# The first 12 entries correspond to month lengths for non-leap years.
+# The remaining 12 entries give month lengths for leap years
+cdef int32_t* days_per_month_array = [
+    31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31,
+    31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
+
+cdef int* sakamoto_arr = [0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4]
+
+# The first 13 entries give the month days elapsed as of the first of month N
+# (or the total number of days in the year for N=13) in non-leap years.
+# The remaining 13 entries give the days elapsed in leap years.
+cdef int32_t* _month_offset = [
+    0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365,
+    0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366]
+
+# ----------------------------------------------------------------------
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef inline int32_t get_days_in_month(int year, Py_ssize_t month) nogil:
+    """Return the number of days in the given month of the given year.
+
+    Parameters
+    ----------
+    year : int
+    month : int
+
+    Returns
+    -------
+    days_in_month : int
+
+    Notes
+    -----
+    Assumes that the arguments are valid.  Passing a month not between 1 and 12
+    risks a segfault.
+    """
+    return days_per_month_array[12 * is_leapyear(year) + month - 1]
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef monthrange(int64_t year, Py_ssize_t month):
+    """
+    Return tuple containing the weekday of the first day of the month and
+    the number of days in the month.
+
+    Parameters
+    ----------
+    year : int
+    month : int
+
+    Returns
+    -------
+    weekday : int
+    days_in_month : int
+
+    Raises
+    ------
+    ValueError if month is invalid
+    """
+    cdef:
+        int32_t days
+
+    if month < 1 or month > 12:
+        raise ValueError("bad month number 0; must be 1-12")
+
+    days = get_days_in_month(year, month)
+    return (dayofweek(year, month, 1), days)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision
+cdef int dayofweek(int y, int m, int d) nogil:
```

Comment:
It seems you are duplicating things from `np_datetime.c` ?

Reply:
>It seems you are duplicating things from np_datetime.c ?

@jorisvandenbossche think porting less than duplicating.  The np_datetime.c version is never actually used in src/datetime (will remove old version in follow-up).  As long as perf doesn't take a hit, the default should by python/cython right?

Reply:
Sorry I don't understand your explanation. Previously it was imported it from np_datetime.c, and you reimplemented it here and changed that import. That seems like duplication rather than porting?

`dayofweek` is maybe not used in np_datetime.c itself, but `is_leapyear` is, so that one cannot just be removed there.

Reply:
> That seems like duplication rather than porting?

You're right.  The intention is to remove the np_datetime.c version in a follow-up PR.

`is_leapyear` _will_ be duplicated because it _is_ used in np_datetime.c, as you noted.  The idea is that np_datetime.c* is a heavy dependency (in practice tends to come with `tslibs.np_datetime` and np_datetime_strings.c) which can be replaced with a much lighter (and more pythonic) `ccalendar` dependency in some cases.

\* The name overlap between np_datetime.c and tslibs.np_datetime is not my favorite.

Reply:
I would rather include extra deps than duplicate code. mark this for a followup.